contracts/cw20-ics20/src/ibc.rs

```diff
@@ -0,0 +1,248 @@
+fn ack_success() -> StdResult<Binary> {
+    let res = Ics20Ack::Result(b"1".into());
+    to_binary(&res)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+/// enforces ordering and versioning constraints
+pub fn ibc_channel_open(
+    _deps: DepsMut,
+    _env: Env,
+    channel: IbcChannel,
+) -> Result<(), ContractError> {
+    enforce_order_and_version(&channel)?;
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+/// record the channel in CHANNEL_INFO
+pub fn ibc_channel_connect(
+    deps: DepsMut,
+    _env: Env,
+    channel: IbcChannel,
+) -> Result<IbcBasicResponse, ContractError> {
+    // we need to check the counter party version in try and ack (sometimes here)
+    enforce_order_and_version(&channel)?;
+
+    let info = ChannelInfo {
+        id: channel.endpoint.channel_id,
+        counterparty_endpoint: channel.counterparty_endpoint,
+        connection_id: channel.connection_id,
+    };
+    CHANNEL_INFO.save(deps.storage, &info.id, &info)?;
+
+    // TODO: add events/attributes here?
+    let res = IbcBasicResponse::default();
+    Ok(res)
+}
+
+fn enforce_order_and_version(channel: &IbcChannel) -> Result<(), ContractError> {
+    if channel.version != ICS20_VERSION {
+        return Err(ContractError::InvalidIbcVersion {
+            version: channel.version.clone(),
+        });
+    }
+    if let Some(version) = &channel.counterparty_version {
+        if version != ICS20_VERSION {
+            return Err(ContractError::InvalidIbcVersion {
+                version: version.clone(),
+            });
+        }
+    }
+    if channel.order != ICS20_ORDERING {
+        return Err(ContractError::OnlyOrderedChannel {});
+    }
+    Ok(())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn ibc_channel_close(
+    _deps: DepsMut,
+    _env: Env,
+    _channel: IbcChannel,
+) -> Result<IbcBasicResponse, ContractError> {
+    // TODO: what to do here?
+    // we will have locked funds that need to be returned somehow
+    unimplemented!();
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+/// Check to see if we have any balance here
+/// We should not return an error if possible, but rather an acknowledgement of failure
+pub fn ibc_packet_receive(
+    deps: DepsMut,
+    _env: Env,
+    packet: IbcPacket,
+) -> Result<IbcReceiveResponse, ContractError> {
+    // TODO: don't let error leak
+    let msg: Ics20Packet = from_binary(&packet.data)?;
+    let channel = packet.src.channel_id;
+    let denom = msg.denom;
+    let amount = Uint128::from(msg.amount);
+    CHANNEL_STATE.update(
+        deps.storage,
+        (&channel, &denom),
+        |orig| -> Result<_, ContractError> {
+            // this will return error if we don't have the funds there to cover the request (or no denom registered)
+            let mut cur = orig.ok_or(ContractError::InsufficientFunds {})?;
+            cur.outstanding = (cur.outstanding - amount)?;
+            Ok(cur)
+        },
+    )?;
```

Comment:
I don't fully understand this block so I may go this wrong. The "no funds" is an IBC application level error that is part of the payload and should not result in an error that reverts the TX in the module.
See https://github.com/cosmos/cosmos-sdk/blob/v0.41.4/proto/ibc/core/channel/v1/channel.proto#L145

Reply:
You are very correct here.
I made a separate function in https://github.com/CosmWasm/cosmwasm-plus/pull/238/commits/0844e34447f03ea464f281e66bb1b0816d8c7aa7 so that I can catch the errors and turn them into "acknowledgement messages containing errors".